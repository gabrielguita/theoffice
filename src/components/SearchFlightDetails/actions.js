import {
  SEARCH_FLIGHT,
  SEARCH_FLIGHT_SUCCESS,
  SEARCH_FLIGHT_FAILURE,
} from "./constants";

export const searchFlight = (payload) => ({
  type: SEARCH_FLIGHT,
  payload,
});

export const searchFlightSuccess = (result) => ({
  type: SEARCH_FLIGHT_SUCCESS,
  payload: result,
});

export const searchFlightFailure = (error) => ({
  type: SEARCH_FLIGHT_FAILURE,
  payload: error,
});
