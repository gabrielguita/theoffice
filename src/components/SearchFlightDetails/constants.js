export const SEARCH_FLIGHT = "SEARCH_FLIGHT";
export const SEARCH_FLIGHT_SUCCESS = "SEARCH_FLIGHT_SUCCESS";
export const SEARCH_FLIGHT_FAILURE = "SEARCH_FLIGHT_FAILURE";
