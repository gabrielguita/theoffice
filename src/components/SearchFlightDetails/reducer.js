import {
  SEARCH_FLIGHT,
  SEARCH_FLIGHT_SUCCESS,
  SEARCH_FLIGHT_FAILURE,
} from "./constants";

export const initialState = {
  flights: {},
  error: null,
  isLoading: false,
};

const flightsReducer = (state = initialState, action) => {
  switch (action.type) {
    case SEARCH_FLIGHT:
      return {
        ...state,
        isLoading: true,
      };
    case SEARCH_FLIGHT_SUCCESS:
      return {
        ...state,
        isLoading: false,
        flights: action.payload,
      };
    case SEARCH_FLIGHT_FAILURE:
      return {
        ...state,
        isLoading: true,
        error: action.payload,
      };

    default:
      return state;
  }
};

export default flightsReducer;
