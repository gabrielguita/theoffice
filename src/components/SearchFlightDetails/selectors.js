import { createSelector } from "reselect";
import get from "lodash/get";

const selectFlight = (state) => state;

const makeSelectFlight = () =>
  createSelector(selectFlight, (substate) => substate);

const makeSelectIsLoading = () =>
  createSelector(selectFlight, (substate) =>
    get(substate, "getFlights.isLoading", null),
  );

const makeSelectError = () =>
  createSelector(selectFlight, (substate) => ({
    error: get(substate, "getFlights.error", null),
    msg: get(substate, "getFlights.flights.message", null),
  }));

const makeSelectFlights = () =>
  createSelector(selectFlight, (substate) => {
    const flights = [];
    const allFlights = get(substate, "getFlights.flights.data", null);

    allFlights &&
      allFlights.map((flight) =>
        flights.push({
          id: flight.id,
          fly_duration: flight.fly_duration,
          flyFrom: flight.flyFrom,
          cityFrom: flight.cityFrom,
          flyTo: flight.flyTo,
          cityTo: flight.cityTo,
          airlines: flight.airlines,
          price: flight.price,
          availability: flight.availability,
          dTimeUTC: flight.dTime,
          aTimeUTC: flight.aTime,
        }),
      );
    return flights;
  });

export {
  makeSelectFlight,
  makeSelectIsLoading,
  makeSelectFlights,
  makeSelectError,
};
