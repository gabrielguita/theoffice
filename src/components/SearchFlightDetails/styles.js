import styled from "styled-components";
import media from "styled-media-query";
import { Button, TextField } from "@material-ui/core";

export const Row = styled.div`
  flex: 1;
`;

export const Seats = styled.span`
  color: ${(props) => (props.value < 4 ? "red" : "green")};
  font-size: 14px;
  margin: 0 7px;
  font-weight: 400;
`;

export const FlightDetails = styled.ul`
  max-width: 500px;
  margin: 0 auto;
  padding: 0;
  li {
    list-style: none;
    color: #000;
    margin-bottom: 20px;
    padding-bottom: 5px;
    border: solid 1px #737373;
    padding: 0px 20px;
  }
  ${media.lessThan("375px")`
    margin: 15px 0 0 0;
  `}
`;

export const SearchButtonContainer = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
  margin: 30px auto 20px auto;
`;

export const SearchButton = styled(Button)`
  display: flex;
  ${media.lessThan("375px")`
    margin: 15px 0 0 0;
  `};
`;

export const Wrapper = styled.div`
  border: solid 1px #ccc;
  padding: 30px 20px;
  border-radius: 15px;
  background: #fdfdfd;
  ${media.lessThan("375px")`
    margin: 15px 0 0 0;
  `}
`;

export const FlightDestination = styled.div`
  flex: 1;
  margin-bottom: 15px;
  ${media.lessThan("375px")`
    margin: 15px 0 0 0;
  `}
`;

export const FlightDuration = styled.div`
  flex: 1;
  margin-bottom: 15px;
  ${media.lessThan("375px")`
    margin: 15px 0 0 0;
  `}
`;

export const FlightFrom = styled(TextField)`
  ${media.lessThan("375px")`
    margin: 15px 0 0 0;
  `}
`;

export const FlightTo = styled(TextField)`
  ${media.lessThan("375px")`
    margin: 15px 0 0 0;
  `}
`;

export const SearchFLight = styled.div`
  width: 100%;
  display: flex;
  ${media.lessThan("375px")`
    margin: 15px 0 0 0;
  `}
`;

export const Picture = styled.img`
  float: left;
  margin-right: 20px;
  max-width: 180px;
  ${media.lessThan("375px")`
    max-width: 100%;
    width: 100%;
    float: none;
    margin: 0 auto;
  `}
`;
