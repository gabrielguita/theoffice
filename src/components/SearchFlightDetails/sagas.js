import { put, all, takeLatest, call } from "redux-saga/effects";
import { searchFlightSuccess, searchFlightFailure } from "./actions";
import { SEARCH_FLIGHT } from "./constants";

import fetchFlightServices from "../../services/searchFlightServices";

export function* searchFlight({ payload }) {
  try {
    const apiGetResponse = yield call(fetchFlightServices, payload);
    yield put(searchFlightSuccess(apiGetResponse));
  } catch (error) {
    yield put(searchFlightFailure(error.message));
  }
}

export default function* rootSagas() {
  yield all([takeLatest(SEARCH_FLIGHT, searchFlight)]);
}
