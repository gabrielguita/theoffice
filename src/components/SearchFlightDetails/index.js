import React, { useState, useEffect } from "react";
import PropTypes from "prop-types";
import moment from "moment";
import { makeStyles } from "@material-ui/core/styles";
import { connect } from "react-redux";
import { Select, InputLabel, MenuItem, FormControl } from "@material-ui/core";
import {
  Wrapper,
  SearchFLight,
  FlightFrom,
  FlightTo,
  SearchButton,
  FlightDestination,
  FlightDuration,
  FlightDetails,
  Row,
  Seats,
  SearchButtonContainer,
} from "./styles";

import {
  makeSelectIsLoading,
  makeSelectFlights,
  makeSelectError,
} from "./selectors";
import { createStructuredSelector } from "reselect";
import { searchFlight as searchFlightAction } from "./actions";
import Spinner from "../../components/Spinner/";
import SearchWeatherDetails from "../SearchWeatherDetails";
import ErrorMessages from "../../components/ErrorMessages";

const useStyles = makeStyles((theme) => ({
  formControl: {
    margin: theme.spacing(1),
    minWidth: 120,
  },
  selectEmpty: {
    marginTop: theme.spacing(2),
  },
}));

const SearchFlightDetails = (props) => {
  const classes = useStyles();
  const [cityFrom, setCityFrom] = useState("");
  const [cityTo, setCityTo] = useState("");
  const [dateFrom, setDateFrom] = useState("");
  const [dateTo, setDateTo] = useState("");
  const [isDisabled, setIsDisabled] = useState(true);
  const { searchFlight, isLoading, flights, hasError } = props;

  const requiredFields =
    cityFrom.trim() && cityTo.trim() && dateFrom.trim() && dateTo.trim();

  useEffect(() => {
    if (requiredFields) {
      setIsDisabled(false);
    } else {
      setIsDisabled(true);
    }
  }, [cityFrom, cityTo, dateFrom, dateTo, requiredFields]);

  const handleOnSubmit = (e) => {
    e.preventDefault();
    searchFlight({
      flyFrom: cityFrom,
      flyTo: cityTo,
      dateFrom: dateFrom,
      dateTo: dateTo,
      v: 3,
    });
  };

  const formatDate = (date) => moment(date).format("DD/MM/YYYY");

  const handleOnChangeCity = (e) =>
    e.target.name === "cityFrom"
      ? setCityFrom(e.target.value)
      : setCityTo(e.target.value);

  const handleChangeDate = (e) =>
    e.target.name === "from"
      ? setDateFrom(formatDate(e.target.value))
      : setDateTo(formatDate(e.target.value));

  return (
    <Wrapper>
      {cityTo && <SearchWeatherDetails data={cityTo} />}
      <form className={classes.container} onSubmit={handleOnSubmit}>
        <SearchFLight>
          <FlightDestination>
            <Row>
              <FormControl className={classes.formControl}>
                <InputLabel shrink id="SelectCityFrom-label">
                  Select city From
                </InputLabel>
                <Select
                  name="cityFrom"
                  labelId="SelectCityFrom-label"
                  id="SelectCityFrom"
                  value={cityFrom}
                  onChange={handleOnChangeCity}
                >
                  <MenuItem value="" disabled>
                    Select city
                  </MenuItem>
                  <MenuItem value={"AMS"}>Amsterdam</MenuItem>
                  <MenuItem value={"BUD"}>Budapest</MenuItem>
                  <MenuItem value={"MAD"}>Madrid</MenuItem>
                </Select>
              </FormControl>
            </Row>
            <Row>
              <FormControl className={classes.formControl}>
                <InputLabel shrink id="SelectCityFrom-label">
                  Select city To
                </InputLabel>
                <Select
                  name="cityTo"
                  labelId="SelectCityFrom-label"
                  id="SelectCityTo"
                  value={cityTo}
                  onChange={handleOnChangeCity}
                >
                  <MenuItem value="" disabled>
                    Select city
                  </MenuItem>
                  <MenuItem value={"AMS"}>Amsterdam</MenuItem>
                  <MenuItem value={"BUD"}>Budapest</MenuItem>
                  <MenuItem value={"MAD"}>Madrid</MenuItem>
                </Select>
              </FormControl>
            </Row>
          </FlightDestination>
          <FlightDuration>
            <Row>
              <FlightFrom
                id="from"
                name="from"
                label="From"
                className={classes.flightFrom}
                onChange={handleChangeDate}
                type="date"
                InputLabelProps={{
                  shrink: true,
                  name: "Flight From",
                }}
              />
            </Row>
            <Row>
              <FlightTo
                name="to"
                id="to"
                label="To"
                className={classes.flightTo}
                onChange={handleChangeDate}
                type="date"
                InputLabelProps={{
                  shrink: true,
                }}
              />
            </Row>
          </FlightDuration>
        </SearchFLight>
        <SearchButtonContainer>
          <SearchButton
            type="submit"
            variant="contained"
            color="primary"
            disabled={isDisabled}
          >
            Search Flight
          </SearchButton>
        </SearchButtonContainer>
        {hasError.msg && <ErrorMessages msg={hasError.msg} />}
      </form>
      {flights && (
        <FlightDetails>
          {flights.map((flight) => (
            <li key={flight.id}>
              <p>From: {flight.cityFrom} </p>
              <p>To: {flight.cityTo} </p>
              <p>
                Departure:{" "}
                {moment.unix(flight.dTimeUTC).format("DD/MM/YYYY HH:mm")}
              </p>
              <p>
                Arrivals:{" "}
                {moment.unix(flight.aTimeUTC).format("DD/MM/YYYY HH:mm")}
              </p>
              <p>
                Airlines companies:
                {flight &&
                  flight.airlines.map((company) => (
                    <small key={company}> {company} </small>
                  ))}
              </p>
              <p>Duration: {flight.fly_duration} </p>
              <p>Price: {flight.price} â‚¬</p>
              {flight.availability.seats && (
                <p>
                  Availability:
                  <Seats value={flight.availability.seats}>
                    {flight.availability.seats}
                  </Seats>
                  seats
                </p>
              )}
            </li>
          ))}
        </FlightDetails>
      )}
      {isLoading && <Spinner />}
    </Wrapper>
  );
};

SearchFlightDetails.propTypes = {
  searchFlightAction: PropTypes.func,
  searchFlight: PropTypes.func,
  isLoading: PropTypes.bool,
  flights: PropTypes.array,
  hasError: PropTypes.object,
};

const mapStateToProps = createStructuredSelector({
  isLoading: makeSelectIsLoading(),
  flights: makeSelectFlights(),
  hasError: makeSelectError(),
});

export function mapDispatchToProps(dispatch) {
  return {
    searchFlight: (data) => dispatch(searchFlightAction(data)),
  };
}

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(SearchFlightDetails);
