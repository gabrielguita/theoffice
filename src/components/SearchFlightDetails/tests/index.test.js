import React from "react";
import SearchFlightDetails from "../index";

describe("<SearchFlightDetails />", () => {
  it("should render correctly", () => {
    expect(<SearchFlightDetails />).toMatchSnapshot();
  });
});
