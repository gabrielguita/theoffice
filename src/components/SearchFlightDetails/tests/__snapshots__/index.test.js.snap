// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchFlightDetails /> should render correctly 1`] = `<UNDEFINED />`;
